### PR TITLE
feat: Support filters in matching engine vector matching

### PR DESCRIPTION
**Pull request by jaycee-li**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-aiplatform/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

This PR adds the feature that was requested in [b/233694218](https://buganizer.corp.google.com/issues/233694218)

### CODE REVIEW

**Comment by nayaknishant** [comment]
Hey @jaycee-li is it necessary to have the Namespace class? I'm wondering what's the easiest way for users to understand how to go about making/using filters.

**Comment by jaycee-li** [comment]
> Hey @jaycee-li is it necessary to have the Namespace class? I'm wondering what's the easiest way for users to understand how to go about making/using filters.

@nayaknishant Yeah I thought about it. But the filter message is defined in gapic, I didn't find a way to make the structure simpler in SDK. If not using Namespace class, users will need to construct the filter (in dictionary or list) by themselves. I feel like this Namespace class can help users better understand the filter message.

**Comment by nayaknishant** [comment]
Sounds good. To emphasize how Namespace is used can we change the example in the comment to `For example, [Namespace("color", ["red"], []), Namespace("shape", [], ["squared"])] will match datapoints that satisfy "red color" and not include datapoints that are "squared"`, or something like that (showing how both `allow_tokens` and `deny_tokens` are used).

**Comment by jaycee-li** [comment]
> Sounds good. To emphasize how Namespace is used can we change the example in the comment to `For example, [Namespace("color", ["red"], []), Namespace("shape", [], ["squared"])] will match datapoints that satisfy "red color" and not include datapoints that are "squared"`, or something like that (showing how both `allow_tokens` and `deny_tokens` are used).

@nayaknishant  Updated.

File: google/cloud/aiplatform/matching_engine/matching_engine_index_endpoint.py [hunk 1]
```diff
@@ -51,6 +51,24 @@ class MatchNeighbor:
...
+class Namespace:
+    """Namespace specifies the rules for determining the datapoints that are eligible for each matching query, overall query is an AND across namespaces.
+    Args:
+        name (str):
+            Required. The name of this Namespace.
+        allow_tokens (List(str)):
+            Optional. The allowed tokens in the namespace.
+        deny_tokens (List(str)):
+            Optional. The denied tokens in the namespace. When a token is denied, then matches will be excluded whenever the other datapoint has that token.
+            For example, if a query specifies {color: red, blue, !purple}, then that query will match datapoints that are red or blue, but if those points are also purple, then they will be excluded even if they are red/blue.
```
Comment (nayaknishant) [inline google/cloud/aiplatform/matching_engine/matching_engine_index_endpoint.py:64]: Let's also change `if a query specifies {color: red, blue, !purple}` to `if a query specifies [Namespace("color", ["red"], ["purple"]), Namespace("color", ["blue], [])]`

Comment (jaycee-li) [reply google/cloud/aiplatform/matching_engine/matching_engine_index_endpoint.py]: Updated

**Review by nayaknishant — APPROVED** [review]

**Pull request merged**